File: apps/ios/Janark/Data/Session/SessionStore.swift
import Foundation
import Combine

struct UiSession: Equatable, Sendable {
    var authenticated: Bool = false
    var anonId: String? = nil
    var hint: String? = nil
}

@MainActor
final class SessionStore: ObservableObject {
    private let defaults: UserDefaults
    private enum Keys {
        static let apiBase = "api_base_url"
        static let theme = "theme_id"
        static let anonId = "anon_id"
        static let hint = "phone_hint"
        static let authenticated = "authenticated"
        static let termsVersion = "accepted_terms_version"
        static let visitorId = "visitor_id"
    }

    static let defaultAPIBaseURL = "http://127.0.0.1:3000"

    @Published var apiBaseURL: String
    @Published var themeId: ThemeId
    @Published var uiSession: UiSession
    @Published var acceptedTermsVersion: String?

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        let storedBase = defaults.string(forKey: Keys.apiBase)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
        var base = (storedBase?.isEmpty == false) ? storedBase! : Self.defaultAPIBaseURL
        while base.hasSuffix("/") { base = String(base.dropLast()) }
        self.apiBaseURL = base
        self.themeId = ThemeId.from(defaults.string(forKey: Keys.theme))
        self.uiSession = UiSession(
            authenticated: defaults.bool(forKey: Keys.authenticated),
            anonId: defaults.string(forKey: Keys.anonId),
            hint: defaults.string(forKey: Keys.hint)
        )
        self.acceptedTermsVersion = defaults.string(forKey: Keys.termsVersion)
    }

    func currentApiBase() -> String {
        var cleaned = apiBaseURL.trimmingCharacters(in: .whitespacesAndNewlines)
        while cleaned.hasSuffix("/") { cleaned = String(cleaned.dropLast()) }
        return cleaned
    }

    func setApiBase(_ url: String) {
        var cleaned = url.trimmingCharacters(in: .whitespacesAndNewlines)
        while cleaned.hasSuffix("/") { cleaned = String(cleaned.dropLast()) }
        defaults.set(cleaned, forKey: Keys.apiBase)
        apiBaseURL = cleaned
    }

    func setTheme(_ id: ThemeId) {
        defaults.set(id.rawValue, forKey: Keys.theme)
        themeId = id
    }

    func setSession(anonId: String?, hint: String?, authenticated: Bool) {
        if let anonId { defaults.set(anonId, forKey: Keys.anonId) }
        else { defaults.removeObject(forKey: Keys.anonId) }
        if let hint { defaults.set(hint, forKey: Keys.hint) }
        else { defaults.removeObject(forKey: Keys.hint) }
        defaults.set(authenticated, forKey: Keys.authenticated)
        uiSession = UiSession(authenticated: authenticated, anonId: anonId, hint: hint)
    }

    func clearSession() {
        defaults.removeObject(forKey: Keys.anonId)
        defaults.removeObject(forKey: Keys.hint)
        defaults.set(false, forKey: Keys.authenticated)
        uiSession = UiSession()
    }

    func acceptTerms(_ version: String) {
        defaults.set(version, forKey: Keys.termsVersion)
        acceptedTermsVersion = version
    }

    @discardableResult
    func visitorId() -> String {
        if let existing = defaults.string(forKey: Keys.visitorId), !existing.isEmpty {
            return existing
        }
        let uuid = UUID().uuidString.replacingOccurrences(of: "-", with: "")
        let id = "vis_" + String(uuid.prefix(24))
        defaults.set(id, forKey: Keys.visitorId)
        return id
    }
}

File: apps/android/app/src/main/java/org/janark/app/data/session/SessionStore.kt
package org.janark.app.data.session

import android.content.Context
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import org.janark.app.BuildConfig
import org.janark.app.ui.theme.ThemeId
import java.util.UUID

private val Context.dataStore by preferencesDataStore("janark_prefs")

data class UiSession(
    val authenticated: Boolean = false,
    val anonId: String? = null,
    val hint: String? = null,
)

class SessionStore(private val context: Context) {
    private object Keys {
        val apiBase = stringPreferencesKey("api_base_url")
        val theme = stringPreferencesKey("theme_id")
        val anonId = stringPreferencesKey("anon_id")
        val hint = stringPreferencesKey("phone_hint")
        val authenticated = booleanPreferencesKey("authenticated")
        val termsVersion = stringPreferencesKey("accepted_terms_version")
        val visitorId = stringPreferencesKey("visitor_id")
    }

    val apiBaseUrl: Flow<String> = context.dataStore.data.map { prefs ->
        prefs[Keys.apiBase]?.trim()?.trimEnd('/')
            ?.takeIf { it.isNotBlank() }
            ?: BuildConfig.DEFAULT_API_BASE_URL.trimEnd('/')
    }

    val themeId: Flow<ThemeId> = context.dataStore.data.map { prefs ->
        ThemeId.from(prefs[Keys.theme])
    }

    val uiSession: Flow<UiSession> = context.dataStore.data.map { prefs ->
        UiSession(
            authenticated = prefs[Keys.authenticated] == true,
            anonId = prefs[Keys.anonId],
            hint = prefs[Keys.hint],
        )
    }

    val acceptedTermsVersion: Flow<String?> = context.dataStore.data.map { it[Keys.termsVersion] }

    suspend fun currentApiBase(): String = apiBaseUrl.first()

    suspend fun setApiBase(url: String) {
        val cleaned = url.trim().trimEnd('/')
        context.dataStore.edit { it[Keys.apiBase] = cleaned }
    }

    suspend fun setTheme(id: ThemeId) {
        context.dataStore.edit { it[Keys.theme] = id.id }
    }

    suspend fun setSession(anonId: String?, hint: String?, authenticated: Boolean) {
        context.dataStore.edit {
            if (anonId != null) it[Keys.anonId] = anonId else it.remove(Keys.anonId)
            if (hint != null) it[Keys.hint] = hint else it.remove(Keys.hint)
            it[Keys.authenticated] = authenticated
        }
    }

    suspend fun clearSession() {
        context.dataStore.edit {
            it.remove(Keys.anonId)
            it.remove(Keys.hint)
            it[Keys.authenticated] = false
        }
    }

    suspend fun acceptTerms(version: String) {
        context.dataStore.edit { it[Keys.termsVersion] = version }
    }

    suspend fun visitorId(): String {
        val existing = context.dataStore.data.first()[Keys.visitorId]
        if (!existing.isNullOrBlank()) return existing
        val id = "vis_${UUID.randomUUID().toString().replace("-", "").take(24)}"
        context.dataStore.edit { it[Keys.visitorId] = id }
        return id
    }
}
